File: backend/app/schemas/poem.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

class PoemBase(BaseModel):
    title: str
    author: str
    dynasty: str
    content: str
    translation: Optional[str] = None
    annotation: Optional[str] = None
    background: Optional[str] = None
    appreciation: Optional[str] = None
    category: Optional[str] = None
    genre: Optional[str] = None
    tags: Optional[str] = None

class PoemCreate(PoemBase):
    pass

class PoemUpdate(BaseModel):
    title: Optional[str] = None
    author: Optional[str] = None
    dynasty: Optional[str] = None
    content: Optional[str] = None
    translation: Optional[str] = None
    annotation: Optional[str] = None
    background: Optional[str] = None
    appreciation: Optional[str] = None
    category: Optional[str] = None
    genre: Optional[str] = None
    tags: Optional[str] = None

class PoemListItem(BaseModel):
    id: int
    title: str
    author: str
    dynasty: str
    content: str
    category: Optional[str] = None
    genre: Optional[str] = None
    view_count: int = 0
    favorite_count: int = 0

    class Config:
        from_attributes = True

class PoemDetail(PoemBase):
    id: int
    view_count: int = 0
    favorite_count: int = 0
    created_at: datetime
    updated_at: datetime
    is_favorited: bool = False

    class Config:
        from_attributes = True

class PoemListResponse(BaseModel):
    items: List[PoemListItem]
    total: int
    page: int
    page_size: int

class PoemSearchParams(BaseModel):
    keyword: str
    search_type: str = Field(default="content", pattern="^(title|author|content)$")

File: frontend/src/api/poem.ts
import request from '@/utils/request'

export interface Poem {
  id: number
  title: string
  author: string
  dynasty: string
  content: string
  category?: string
  genre?: string
  view_count: number
  favorite_count: number
}

export interface PoemDetail extends Poem {
  translation?: string
  annotation?: string
  background?: string
  appreciation?: string
  tags?: string
  created_at: string
  updated_at: string
  is_favorited: boolean
}

export interface PoemListResponse {
  items: Poem[]
  total: number
  page: number
  page_size: number
}

export const poemApi = {
  getList(params: {
    page?: number
    page_size?: number
    dynasty?: string
    author?: string
    category?: string
    genre?: string
  }) {
    return request.get<PoemListResponse>('/poems', { params })
  },

  getDetail(id: number) {
    return request.get<PoemDetail>(`/poems/${id}`)
  },

  search(params: {
    keyword: string
    search_type?: 'title' | 'author' | 'content'
    page?: number
    page_size?: number
  }) {
    return request.get<PoemListResponse>('/poems/search', { params })
  },

  getRandom(dynasty?: string) {
    return request.get<PoemDetail>('/poems/random', {
      params: dynasty ? { dynasty } : {}
    })
  },

  incrementView(id: number) {
    return request.post(`/poems/${id}/view`)
  },

  favorite(id: number) {
    return request.post(`/poems/${id}/favorite`)
  },

  unfavorite(id: number) {
    return request.delete(`/poems/${id}/favorite`)
  }
}
